### PR TITLE
add reporter

### PR DESCRIPTION
**Pull request by oibe**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -15,6 +15,7 @@
     "lodash": "^4.15.0",
     "my-local-ip": "^1.0.0",
     "node-int64": "^0.4.0",
+    "thriftify": "^1.0.0-alpha14",
```
Comment (yurishkuro) [inline package.json]: what's the difference with thriftrw?

File: src/decls/types.js [hunk 2]
```diff
@@ -63,3 +63,21 @@ declare type startSpanArgs = {
...
+    flush(): SenderResponse;
+    close(): void;
+};
+
+declare class Logger {
+    info(msg: string): void;
+    debug(msg: string): void;
+    warn(msg: string): void;
+    error(msg: string): void;
+}
```
Comment (yurishkuro) [inline src/decls/types.js]: wait, so before you were creating whole new fils, with licenses, for each of the reporter interface, but now the logger doesn't get its own file?

File: src/reporter.thrift [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+namespace java com.uber.jaeger.thriftjava
+
+// TODO(oibe) remove this file, also thriftify does not support unions
```
Comment (yurishkuro) [inline src/reporter.thrift]: ?

File: src/reporter.thrift [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+// TagValue is a union of different types representing a strongly typed value of a tag.
+// TODO: in Go all these fields become pointers, causing lots unnecessary memory allocs,
+// so we may want to revisit and maybe just use a byte array with enum directly in Tag.
+struct TagValue {
```
Comment (yurishkuro) [inline src/reporter.thrift]: this is the wrong thrift idl

File: src/reporters/remote_reporter.js [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+export default class RemoteReporter {
+    _bufferFlushInterval: number;
+    _logger: Logger;
+    _sender: Transport;
+    _intervalHandle: any;
+
+    constructor(sender: Transport,
+                options: any = {}) {
+        if (!options.bufferFlushInterval) {
```
Comment (yurishkuro) [inline src/reporters/remote_reporter.js]: I don't think you need the if

File: src/reporters/remote_reporter.js [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+            this._bufferFlushInterval = options.bufferFlushInterval || DEFAULT_BUFFER_FLUSH_INTERVAL_MILLIS;
+        }
+        this._logger = options.logger || NullLogger;
```
Comment (yurishkuro) [inline src/reporters/remote_reporter.js]: or `NullLogger()`?

File: src/reporters/remote_reporter.js [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        this._sender = sender;
+        this._intervalHandle = setInterval(() => {
+            this.flush();
+        }, this._bufferFlushInterval);
+    }
+
+    report(span: Span): void {
+        let response: SenderResponse = this._sender.append(span._toJSON());
```
Comment (yurishkuro) [inline src/reporters/remote_reporter.js]: toJSON?

File: src/reporters/remote_reporter.js [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import NullLogger from '../logger.js';
+
+const DEFAULT_QUEUE_SIZE = 100
```
Comment (yurishkuro) [inline src/reporters/remote_reporter.js]: which queue?

File: src/reporters/udp_sender.js [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+        }
+
+        let thriftBuffer: Buffer = thriftify.toBuffer({spans: this._spanBuffer}, this._spec, 'Agent::emitZipkinBatch_args');
+        this._client.send(thriftBuffer, 0, thriftBuffer.length, PORT, HOST);
+
+        this._reset();
+
+        if (callback) {
+            callback();
+        }
```
Comment (yurishkuro) [inline src/reporters/udp_sender.js]: that's an odd design, combining async callback with the actual return value.

File: src/span.js [hunk 10]
```diff
@@ -26,9 +26,9 @@ import NullLogger from './logger.js';
 
 export default class Span {
     _tracer: any;
-    _name: string;
+    _operationName: string;
     _spanContext: SpanContext;
-    _start: number;
+    _timestamp: number;
```
Comment (yurishkuro) [inline src/span.js]: that's a bad design on my part, we should rename to startTime in the IDL

File: src/reporter.thrift (hunk 3 above)
Comment (oibe) [inline src/reporter.thrift]: This is a temporary thrift for testing serialization.  I'll sync with Rooz, and get the real definition.

File: src/reporters/remote_reporter.js (hunk 7 above)
Comment (oibe) [inline src/reporters/remote_reporter.js]: poorly named.  Should be more like toThrift.

File: src/reporters/udp_sender.js (hunk 9 above)
Comment (oibe) [inline src/reporters/udp_sender.js]: its for testing purposes.

Comment (yurishkuro) [inline src/reporters/udp_sender.js]: if this whole function is synchronous, why does it need an async callback?

File: src/_interfaces/reporter.js [hunk 11]
```diff
@@ -23,6 +23,17 @@ import Span from '../span.js';
...
+    close(callback: ?Function): void;
+};
+
+declare class Transport {
+    append(span: Span): SenderResponse;
+    flush(): SenderResponse;
+    close(): void;
+};
+
+declare type SenderResponse = {
```
Comment (yurishkuro) [inline src/_interfaces/reporter.js]: name mismatch: TransportResponse

File: src/reporters/udp_sender.js [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+import {Thrift} from 'thriftrw';
+import Span from '../span.js';
+
+const HOST = 'localhost';
+const PORT =  5775;
+const DEFAULT_UDP_SPAN_SERVER_HOST_PORT = `${HOST}:${PORT}`;
+const EMIT_SPAN_BATCH_OVERHEAD = 30;
+const UDP_PACKET_MAX_LENGTH = 65000;
+
+export default class UDPSender {
```
Comment (yurishkuro) [inline src/reporters/udp_sender.js]: name mismatch - let's settle on Sender (and SenderResult), not Transport

File: src/_interfaces/reporter.js [hunk 13]
```diff
@@ -23,6 +23,17 @@ import Span from '../span.js';
...
 declare interface Reporter {
     report(span: Span): void;
-    flush(callback: Function): void;
-    close(callback: Function): void;
-}
+    flush(callback: ?Function): void;
+    close(callback: ?Function): void;
+};
+
+declare class Transport {
```
Comment (yurishkuro) [inline src/_interfaces/reporter.js]: Sender

File: src/span.js (hunk 10 above)
Comment (yurishkuro) [inline src/span.js]: it's been renamed

File: src/span.js [hunk 14]
```diff
@@ -202,4 +202,35 @@ export default class Span {
...
+            let tag = this._tags[i];
+            tags.push({
+                key: tag.key,
+                tagType: 'STRING', // TODO(oibe) support multiple tag types
+                vStr: tag.value,
+                vDouble: 0,
+                vBool: false,
+                vInt16: 0,
+                vInt32: 0,
+                vInt64: 0,
```
Comment (yurishkuro) [inline src/span.js]: we've changed these to optional and removed all ints but int64

**Comment by yurishkuro** [comment]
LGTM

**Pull request merged**